src/python/pants/core/goals/tailor.py

```diff
@@ -117,9 +117,11 @@ def for_target_type(
         comments: Iterable[str] = tuple(),
         build_file_name: str = "BUILD",
     ):
-        owned_sources = (
-            (kwargs or {}).get("sources") or default_sources_for_target_type(target_type) or tuple()
-        )
+        maybe_owned_sources = (kwargs or {}).get("sources")
+        default_sources = default_sources_for_target_type(target_type)
+        if (maybe_owned_sources or triggering_sources) and default_sources is None:
+            raise ValueError(f"Target type {target_type.__name__} does not have a sources field.")
+        owned_sources = maybe_owned_sources or default_sources or tuple()
```

Comment:
cc @benjyw: I had to introduce this work-around in order to allow `tailor` to work with a target type that does not have a `Sources` field, i.e. `GoExternalModule`. 

This change will be made in a separate PR.

Reply:
broken out in https://github.com/pantsbuild/pants/pull/12409